### ISSUE TITLE
Move to enum-based texturing calls, instead of text based.

### ISSUE THREAD

**Issue by johnkslang** [KhronosGroup/glslang#8]
Would like to move to enum-based operators for texturing and image operations.

Currently, almost all built-in functions are mapped to the TOperator enum, and can be dealt with numerically.  However, this is not true for texturing and imaging operations.  It would be better to do all this consistently and eliminate text-based comparisons.

As consumers of glslang would break and need to adapt to this, this is a place to gather input.  (As is the Khronos public bug 1346.)

See 

```
const bool PureOperatorBuiltins = false;  // could break backward compatibility; pending feedback
```

in Initialize.cpp, where this is currently set up, but turned off.

**Comment by johnkslang** [comment]
Internal support for this, meaning within the front-end (GLSL -> AST), is complete now.  It is still turned off, but will turn on soon, so no consumer needs to do text compares of built-in function names; they will all work with enumerants (TOperator) instead.

If turning this on breaks things, it will be trivial to manually switch it back, by setting 

```
PureOperatorBuiltins = false
```

The idea is that, over time, only the "correct" way of being enumerant-based will be maintained, as it is not good to have a mix of enumerants and text comparisons for built-in functions.  (Original idea was a subset was backed by a library, rather than ops, but that's ancient history; they're all ops.)

**Comment by johnkslang** [comment]
This is now in, at fc51d284aae35e6ab6f15b1439eff7b674214165.

**Comment by mawww** [comment]
Hello,

With that change, would not it make sense to disable unary conversion for builtin function calls that have a single argument ? A strange consequence of the current behaviour is that `textureSize(mySamplerBuffer)` is an TIntermUnary and `textureSize(mySampler2D, lod)` is a TIntermAggregate.

It seems more consistent to have builtins that use function call syntax all be TIntermAggregate.

Cheers,

Maxime Coste.

**Comment by mawww** [comment]
Ping,

for AST consumers, having to deal with texturing functions both in a unary and aggregate intermediate is a pain, and does not make much sense conceptually (IMHO). I would expect unary to be used for unary operators.

**Comment by johnkslang** [comment]
@mawww Agreed.  The issue is breaking existing use.  I can see it for texturing, as it's a new thing.  Are you wanting this change to be made for all built-in operations, or just the texturing-like operations?  It does seem sensible for sin() to be unary, and things are working that way for a long time, but I still agree it would be more convenient of a larger class came in as aggregate.

**Comment by mawww** [comment]
Hello,

I do not have a strong opinion regarding sin/cos like operations, they could be either TIntermUnary or TIntermAggregate, as long as its consistent and predictible (I'd like to be able to look at a TOperator value and easily determine if its going to be an unary or aggregate).

I'd say constructors and texture/images functions should be all aggregates.

More tricky cases are aparametric operators (barriers for example), they are aggregates, which is a little surprising related to single parameter ops (like sin) (0-params -> aggregate, 1-param -> unary, 2+-param -> aggregate again).

So ultimately, a possibility could be just to say "if we use a function syntax, its an aggregate, if we use an operator syntax, its a TIntermUnary/Binary". However that would be an intrusive change.

To recap, I am strongly for texture/constructors to be all aggregates, slightly for all function like operations to be aggregates.

**Comment by johnkslang** [comment]
Without too much intrusion, we could possibly move all image/texture ops to be an aggregate, since this is very recent to switch to enumerant-based image/texture ops.

More generally, it might work as a mode to request everything be an aggregate.  Note though unary nodes are much smaller and more efficient than aggregate nodes, and there are lots of builtins that act like unary operators, so maybe it's not really the right thing to do.  (Operations with 0 operands are comparatively rare.)

Other input "yes" or "no" on changing all image/texturing ops to always be aggregate nodes?  I would like to do it, to get all such related ops handled in the same part of the code.

**Comment by arcady-lunarg** [comment]
The specific issue here has been addressed by the commit mentioned above. Any further discussion of potential changes to the AST should go in a separate issue.
